### PR TITLE
Use SSHJ for SSH public key authentication

### LINKED ISSUE #448: Support for modern SSH Ciphers, Keyex and MACs
Hi thanks for taking the time to make this project, I use all the time.

In regards to SSH I think all users will benefit by the support of modern ssh standards.
Specifically:
KexAlgorithms curve25519-sha256@libssh.org,diffie-hellman-group-exchange-sha256
Ciphers chacha20-poly1305@openssh.com,aes256-gcm@openssh.com,aes256-ctr
MACs hmac-sha2-512-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-256

### LINKED ISSUE #494: StrictHostKeyChecking enables MITM attacks
Password store does not validate the connected ssh server (it disables StrictHostKeyChecking), so a man in the middle can spoof the server (e.g. using https://github.com/jtesta/ssh-mitm). This can be used to steal username/password, key names and encrypted key contents.

### PR DESCRIPTION
**Pull request by FabianHenneke**
## :loudspeaker: Type of change

- [x] Bugfix
- [x] New feature
- [x] Enhancement
- [x] Refactoring

## :scroll: Description

Add an SSHJ backend for SSH public key and password authentication and enable it for public key authentication.

More specifically, this commit achieves the following:

- Adds a JGit SSH backend based on SSHJ, which is much more modern and well-designed than Jsch. As a result, we can now use, configure and add modern key types/ciphers/MACs.
- Enable the new backend for SSH sessions using public key authentication.
- Performs host key verification via a simple trust-on-first-use model: The host key encountered on the first connection is stored in `.host_key` and compared on all future connection attempts.
- GitAsyncTask is translated to Kotlin, refactored and cleaned up, which improves error messages and should simplify a future rewrite with coroutines.

There is currently a minor usability issue: If the user imports a private key in OpenSSH format (rather than the old-style PKCS8 format), the connection will fail if an incorrect passphrase is entered, there is no second passphrase prompt. This is fixed by https://github.com/hierynomus/sshj/pull/587.

## :bulb: Motivation and Context

Fixes #494 and fixes #448 for SSH public key authentication, which is by far the most common. I expect us to transition over to SSHJ for password authentication rather soon. OpenKeychain integration may take more time, so we could unlink the issues if you prefer it that way.

## :green_heart: How did you test it?
I verified that sync/pull/push work with both RSA keys generated in the app as well as imported ssh-ed25519 keys. I also verified that the host key verification is effective.

Please do test with all kinds of SSH keys and server configs, as a new backend could always lead to regressions. 

## :pencil: Checklist

- [x] I formatted the code with the IDE's reformat action (Ctrl + Shift + L/Cmd + Shift + L)
- [x] I reviewed submitted code

## :crystal_ball: Next steps

- Use the new backend also for SSH password authentication.
- When https://github.com/hierynomus/sshj/pull/586 gets merged, explore the possibility of using the Android Keystore instead of simple files for the storage of SSH private keys.
- Replace the current SSH keygen flow relying on Jsch with a new flow based on an Android Keystore key.
- Switch OpenKeychain authentication to the new backend. This also gets easier with https://github.com/hierynomus/sshj/pull/586, but could probably also be done without it.

## :camera_flash: Screenshots / GIFs

Example of failed host key verification:
Screenshot_20200527-094002

### CODE REVIEW

**Comment by FabianHenneke** [comment]
The CI fails should now be fixed, there was a lint issue triggered by BC.

**Comment by msfjarvis** [comment]
Awesome stuff, great job on getting this done so quickly! I'll test and review this in a few hours.

File: app/proguard-rules.pro [hunk 1]
```diff
@@ -24,3 +24,6 @@
 -dontobfuscate
 -keep class com.jcraft.jsch.**
```
Comment (msfjarvis) [inline app/proguard-rules.pro:25]: Just a note, this can go

Comment (FabianHenneke) [reply app/proguard-rules.pro:25]: Not yet, unfortunately, unless I missed something. This commit does not fully replace `Jsch`, we will have two SSH backends in parallel for some time.

**Comment by FabianHenneke** [comment]
Please don't merge this after the review, it might be possible to improve the BouncyCastle integration and I am still looking into that.

File: app/proguard-rules.pro (hunk 1 above)
Comment (msfjarvis) [reply app/proguard-rules.pro:25]: And that's what I get for reading two files on the go

**Comment by Skrilltrax** [comment]
@msfjarvis this good enough to start working on top of it ?

**Comment by msfjarvis** [comment]
> @msfjarvis this good enough to start working on top of it or will there be more changes required to it first ?

Fabian's still working on it and we need to battle test it in production before the Jsch backend can be phased out. It's a long process.

**Comment by FabianHenneke** [comment]
Okay, this PR should be ready for review now. The BouncyCastle setup may not be ideal yet with respect to performance, but we can always revisit that later. It did perform pretty well in my tests.

@msfjarvis I rebased onto the current master, but when I restart Password Store, I just see a blank activity. Is it possible that something is amiss with the ViewBinding commits?

**Comment by msfjarvis** [comment]
> Okay, this PR should be ready for review now. The BouncyCastle setup may not be ideal yet with respect to performance, but we can always revisit that later. It did perform pretty well in my tests.

:+1: 

> @msfjarvis I rebased onto the current master, but when I restart Password Store, I just see a blank activity. Is it possible that something is amiss with the ViewBinding commits?

Yeah I'm looking into it...

**Comment by msfjarvis** [comment]
Fixed that, PR coming up

**Comment by FabianHenneke** [comment]
I ended up making one more change which moves the BC setup into Application and hopefully chooses the priorities of the security provider optimally. But who knows...

**Comment by msfjarvis** [comment]
> I ended up making one more change which moves the BC setup into Application and hopefully chooses the priorities of the security provider optimally. But who knows...

The only crypto operations this app runs are these ones that we just introduced. Everything else is done by OpenKeychain who maintain their own bouncycastle fork which runs in their own app completely uncoupled from us. We should be fine.

**Comment by FabianHenneke** [comment]
> > I ended up making one more change which moves the BC setup into Application and hopefully chooses the priorities of the security provider optimally. But who knows...
> 
> The only crypto operations this app runs are these ones that we just introduced. Everything else is done by OpenKeychain who maintain their own bouncycastle fork which runs in their own app completely uncoupled from us. We should be fine.
> 
> And well, Tink has it's own thing for `androidx.security` which again isn't relying on Android.

That is one aspect of it, but I also want to make sure that we use the native implementations of algorithms whenever possible since they are likely faster. The code in its current form should insert the Java BC provider at the same position the Android BC provider was at, which seems like the safest bet.

**Comment by msfjarvis** [comment]
Seems to work correctly when upgrading but I am having trouble getting a clone going on a fresh install (most probably unrelated to this) so I'll just review this based on the stuff I've tested so far and will look into the bug separately tomorrow.

File: app/src/main/java/com/zeapo/pwdstore/git/GitAsyncTask.kt [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+            setCancelable(false)
+            show()
+        }
+    }
+
+    override fun doInBackground(vararg commands: GitCommand<*>): Result? {
+        var nbChanges: Int? = null
+        for (command in commands) {
+            try {
+                if (command is StatusCommand) {
```
Comment (msfjarvis) [inline app/src/main/java/com/zeapo/pwdstore/git/GitAsyncTask.kt:56]: worthwhile making this a `when` clause? IMO it'd be a bit more readable

Comment (FabianHenneke) [reply app/src/main/java/com/zeapo/pwdstore/git/GitAsyncTask.kt]: Definitely, done.

**Comment by FabianHenneke** [comment]
We can merge this now or wait until you had time to reproduce/resolve that bug. As you prefer.

I have received a review for one of the upstream PRs, so this might work as intended quite soon.

**Comment by msfjarvis** [comment]
> We can merge this now or wait until you had time to reproduce/resolve that bug. As you prefer.

I want to confirm the PR isn't causing the regression (unlikely but just wanna be sure) so I'm gonna merge this tomorrow when I triage the bug.

> I have received a review for one of the upstream PRs, so this might work as intended quite soon.

Awesome 👍

**Comment by msfjarvis** [comment]
Can't repro the bug now...

**Pull request merged**